--- a/qmi_wwan_q.c
+++ b/qmi_wwan_q.c
@@ -828,7 +828,7 @@ static struct rtnl_link_stats64 *_rmnet_
 		stats64 = per_cpu_ptr(dev->stats64, cpu);
 
 		do {
-			start = u64_stats_fetch_begin_irq(&stats64->syncp);
+			start = u64_stats_fetch_begin(&stats64->syncp);
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5,18,0)
 			rx_packets = stats64->rx_packets;
 			rx_bytes = stats64->rx_bytes;
@@ -840,7 +840,7 @@ static struct rtnl_link_stats64 *_rmnet_
 			tx_packets = u64_stats_read(&stats64->tx_packets);
 			tx_bytes = u64_stats_read(&stats64->tx_bytes);
 #endif
-		} while (u64_stats_fetch_retry_irq(&stats64->syncp, start));
+		} while (u64_stats_fetch_retry(&stats64->syncp, start));
 
 
 		stats->rx_packets += rx_packets;
@@ -1295,11 +1295,7 @@ static int qmap_register_device(sQmiWwan
     priv->dev = pDev->mpNetDev;
     priv->qmap_version = pDev->qmap_version;
     priv->mux_id = QUECTEL_QMAP_MUX_ID + offset_id;
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 17, 0)
-    memcpy (qmap_net->dev_addr, real_dev->dev_addr, ETH_ALEN);
-#else
-    eth_hw_addr_set (real_dev, qmap_net->dev_addr);
-#endif
+    dev_addr_set(qmap_net, real_dev->dev_addr);
 
 #ifdef QUECTEL_BRIDGE_MODE
 	priv->bridge_mode = !!(pDev->bridge_mode & BIT(offset_id));
