--- a/scripts/netifd-wireless.sh
+++ b/scripts/netifd-wireless.sh
@@ -204,6 +204,43 @@ _wdev_wrapper \
 	wireless_process_kill_all \
 	wireless_set_retry \
 
+wireless_vif_parse_encryption_rsno() {
+	json_get_vars encryption_rsno encryption_rsno_2
+	set_default encryption_rsno none
+	set_default encryption_rsno_2 none
+
+	rsno_auth_type=none
+	rsno_auth_type_2=none
+	rsno_wpa_cipher="CCMP"
+	rsno_wpa_cipher_2="GCMP-256"
+
+	case "$encryption_rsno" in
+		*ccmp256) rsno_wpa_cipher="CCMP-256";;
+		*aes|*ccmp) rsno_wpa_cipher="CCMP";;
+		*gcmp256) rsno_wpa_cipher="GCMP-256";;
+		*gcmp) rsno_wpa_cipher="GCMP";;
+	esac
+
+	case "$encryption_rsno_2" in
+		*ccmp256) rsno_wpa_cipher_2="CCMP-256";;
+		*aes|*ccmp) rsno_wpa_cipher_2="CCMP";;
+		*gcmp256) rsno_wpa_cipher_2="GCMP-256";;
+		*gcmp) rsno_wpa_cipher_2="GCMP";;
+	esac
+
+	case "$encryption_rsno" in
+		psk3*|sae*)
+			rsno_auth_type=sae
+		;;
+	esac
+
+	case "$encryption_rsno_2" in
+		psk3*|sae*)
+			rsno_auth_type_2=sae
+		;;
+	esac
+}
+
 wireless_vif_parse_encryption() {
 	json_get_vars encryption
 	set_default encryption none
@@ -214,6 +251,10 @@ wireless_vif_parse_encryption() {
 
 	if [ "$hwmode" = "ad" ]; then
 		wpa_cipher="GCMP"
+	elif [ "$_w_mode" = "sta" ]; then
+		wpa_cipher="CCMP CCMP-256 GCMP GCMP-256"
+	elif [ "$encryption" == "sae-ext" ] ;then
+		wpa_cipher="GCMP-256"
 	else
 		wpa_cipher="CCMP"
 	fi
@@ -223,6 +264,8 @@ wireless_vif_parse_encryption() {
 		*gcmp256) wpa_cipher="GCMP-256";;
 		*gcmp) wpa_cipher="GCMP";;
 		wpa3-192*) wpa_cipher="GCMP-256";;
+		sae_sae-ext) wpa_cipher="CCMP GCMP-256";;
+		sae-ext-mixed) wpa_cipher="CCMP GCMP-256";;
 	esac
 
 	# 802.11n requires CCMP for WPA
@@ -269,6 +311,12 @@ wireless_vif_parse_encryption() {
 		psk3-mixed*|sae-mixed*)
 			auth_type=psk-sae
 		;;
+		sae-ext)
+			auth_type=sae-ext
+		;;
+		sae-ext-mixed*)
+			auth_type=psk-sae-ext
+		;;
 		psk3*|sae*)
 			auth_type=sae
 		;;
@@ -378,6 +427,7 @@ _wdev_common_iface_config() {
 
 _wdev_common_iface_config() {
 	config_add_string mode ssid encryption 'key:wpakey'
+	config_add_string encryption_rsno encryption_rsno_2
 	config_add_boolean bridge_isolate
 	config_add_array tags
 }
